# Monty Hall

#### The Monty Hall problem was a TV show in which the host had a guest participant pick one of three doors in an attempt to win a car (which is behind only one of the doors). After the participant chooses the door, the host opens one of the other 2 doors and shows the guest it has nothing behind it. Then, the participant is asked whether he would like to switch doors. This simulation aims to find an answer to that question, to find out if the chances of winning the car are bigger if the participant switches the door or not. Here, we are assuming that the participant always switches doors. The scenario is played one million times and the number of wins and losses is counted. The final result is the win and loss ratios.

In [1]:
import random

# Number of simulations
games = 1000000

# Number of wins and losses
win = 0
loss = 0

# For each simulation
for attempt in range(games):
    
    # Define doors names and define a random door for the participant and another for the prize
    doors = ['A','B','C']
    participant = random.choice(doors)
    prize = random.choice(doors)
    
    # If the participant and prize doors are the same
    if participant == prize:
        # Select the doors that are different than the participant/prize door
        rest = list(set(participant) ^ set(doors))
        # Remove a random door from the remaining doors
        remove = random.choice(rest)
        doors.remove(remove)
        # The participant switches doors
        participant = list(set(participant) ^ set(doors))[0]
    
    # If the participant and prize doors are NOT the same
    else:
        # Select the doors that are different than the participant and prize doors
        rest = list(set(doors) ^ set(list(participant) + list(prize)))
        # Remove a random door from the remaining doors
        remove = random.choice(rest)
        doors.remove(remove)
        # The participant switches doors
        participant = list(set(participant) ^ set(doors))[0]
    
    # Towards the end of the simulation, if the participant chose the prize door, win is incremented by one.
    if participant == prize:
        win += 1
    # If the participant did not choose the prize door, loss is incremented by one.
    else:
        loss += 1
        
# Print Win and Loss percentages.
print(win/games, loss/games)

0.666932 0.333068


#### As seen in the simulation above, the participant wins 66.69% of the times if the offer to switch doors is taken. Therefore, switching is better than keeping the initial choice!